providers/dynamodb/shedlock-provider-dynamodb/src/main/java/net/javacrumbs/shedlock/provider/dynamodb/DynamodbLockProvider.java

```diff
@@ -0,0 +1,306 @@
+                                    .withLong(":" + LOCK_UNTIL, lockConfiguration.getLockAtMostUntil().toEpochMilli())
+                                    .withLong(":" + LOCKED_AT, Instant.now().toEpochMilli()))
+                    .withReturnValues(ReturnValue.ALL_OLD);
+
+            UpdateItemOutcome updateItemOutcome = table.updateItem(spec);
+            return (updateItemOutcome.getItem() == null);
+        }
+
+        private Map<String, AttributeValue> findLockRecord(LockConfiguration lockConfiguration) {
+            DynamoDB dynamoDB = new DynamoDB(amazonDynamoDB);
+            Table table = dynamoDB.getTable(tableName);
+
+            GetItemSpec spec = new GetItemSpec()
+                    .withPrimaryKey(NAME, lockConfiguration.getName());
+
+            return table.getItemOutcome(spec).getGetItemResult().getItem();
+        }
+
+        private void createTableIfNotExists() throws InterruptedException {
+            if (!isTableExist(tableName)) {
+                createTable(tableName);
+            }
+        }
+
+        private boolean isTableExist(String tableName) {
+
+            // If table has been created already and we are aware,
+            // then skip the expensive operations
+            if (IS_TABLE_EXIST) {
+                return true;
+            }
+
+            // If we are not aware of the table's existance
+            // then find out
+            boolean moreTablesToScan = true;
+            boolean isTableExist = false;
+            String lastTableName = null;
+
+            ListTablesRequest listTablesRequest;
+            while (moreTablesToScan) {
+                if (lastTableName != null) {
+                    listTablesRequest = new ListTablesRequest();
+                } else {
+                    listTablesRequest = new ListTablesRequest()
+                            .withExclusiveStartTableName(lastTableName);
+                }
+
+                ListTablesResult tableListResult = amazonDynamoDB.listTables(listTablesRequest);
+                List<String> tableNameList = tableListResult.getTableNames();
+
+                if (tableNameList.isEmpty()) {
+                    //Table does not exist
+                    moreTablesToScan = false;
+                    isTableExist = false;
+                } else {
+                    isTableExist = tableNameList.stream().anyMatch(t -> t.equalsIgnoreCase(tableName));
+                    if (isTableExist) {
+                        moreTablesToScan = false;
+                    }
+                }
+
+                lastTableName = tableListResult.getLastEvaluatedTableName();
+                if (lastTableName == null) {
+                    moreTablesToScan = false;
+                }
+
+            }
+
+            IS_TABLE_EXIST = isTableExist;
+            return isTableExist;
+        }
+
+        private void createTable(String tableName) throws InterruptedException {
```

Comment:
What about moving this code out of the LockProvider? Either leave the table creation completely on the user (as we do for example in JDBC) or provide a utility method which has to be explicitly called before this LockProvider is used.

On one hand it would make the usage more complicated, on the other hand would decouple the code and force the user to think about quotas and such.

Reply:
Right now, there are two constructors. 
- public DynamodbLockProvider(AmazonDynamoDB amazonDynamoDB)
- public DynamodbLockProvider(AmazonDynamoDB amazonDynamoDB, String tableName, Duration minLockTime, long concurrentProcessCount)

In former, user needs to specify the handle to dynamo db that they want to use. Table name and the through put us selected as a default. This is the easiest option. 

In the second user needs to specify handle to dynamo db they want to use and in addition to that, name of the table and some parameters around their anticipated use of the locking framework which is then in turn to determine throughput they may require. This gives a little more control to the user in case, they want to use their own table name and know about the locking and unlocking frequency. 

This can be taken to the next level of a power user where they might like to provision tables first and then use this framework to configure with the existing table. This might be suitable for bigger teams where separate group owns and does the infrastructure provisioning work like table creation, bandwidth allocation, etc. The application team would just build the application. Although this can already be achieved using the 2nd constructor if the table already exists; or with the first constructor where the table name happens to be "shedLock", the default name. 

For neatness, a 3rd constructor can be offered - 
- public DynamodbLockProvider(AmazonDynamoDB amazonDynamoDB, String tableName)
This will utilize the already created table; but might create the table with default throughput if does not exist already.
Or, a 4th constructor
public DynamodbLockProvider(AmazonDynamoDB amazonDynamoDB, Table table)
This would provide even greater control to user to even programmatically or manually create a table and pass it on.

To make it even user friendly, we can wrap all these variation with a Builder which can guide the user to the right choice. 

Thoughts?

Reply:
My concern is more about the design. A class should ideally do only one thing. A LockProvider should just provide a lock and nothing more. Table creation should be concern of some other class. It's not just a theoretical concern, I expect that there will be lot's of different ways how people create their tables and I do not want to prefer one of them before the other. So my favorite solution would be a LockProvider with only one constructor taking Table as an argument and the rest of the code in some other class. May be builder, utility class or what you deem fit. Thanks.

Reply:
SRP certainly makes sense and that allows user to create the table in their desired way